### ISSUE TITLE
Regression with pattern matching for Sum types

### ISSUE THREAD

**Issue by omalaspinas** [diku-dk/futhark#1218]
Hello,
I think there may have been a regression with the pattern matching for sum types. I think the checking of all tuple variants has a problem.

While in futhark 18.5 the following code does not compile (rightfully):

```futhark
type t = #a | #b

let foo (a: t) (b: t) = 
    match (a, b)
    case (#a, #a) -> 1f64
    case (#b, #b) -> 2f64

let bar (a: t) (b: t) = 
    match (a, b)
    case (#a, #b) -> 1f64
    case (#b, #a) -> 1f64
```

It does in the latest git version. The only correct version for the compiler should have been something I think:

```futhark
type t = #a | #b

let baz (a: t) (b: t) = 
    match (a, b)
    case (#a, #a) -> 1f64
    case (#b, #b) -> 1f64
    case (#a, #b) -> 1f64
    case (#b, #a) -> 1f64
```

Edit: added "correct" version.

**Comment by athas** [comment]
Odd, I don't recall touching this since 18.5.

**Comment by athas** [comment]
I can't reproduce.  I get a type error with the newest version of the compiler.

**Comment by omalaspinas** [comment]
I just cloned the repo, stack build --fast / install and I have no errors... Could --fast be the culprit?

**Comment by athas** [comment]
What does `futhark --version` say?  Are you sure you're on the right branch?  This used to be a bug, but it was fixed in 0.18.2.

**Comment by omalaspinas** [comment]
I cleaned the stack cache and rebuilt. The error is gone. Sorry for the issue. I close.
